src/main/jbake/content/blog/2016/12/19/hawkular-metrics-compression.adoc

```diff
@@ -0,0 +1,93 @@
+= Hawkular-Metrics and the new compression algorithm
+Michael Burman
+2016-12-19
+:jbake-type: post
+:jbake-status: published
+:jbake-tags: blog, metrics, compression, gorilla, storage, performance
+
+In Hawkular-Metrics release 0.20.0 we introduced a new compression capability that reduces the required storage space. This blog post will describe details about the implementation in the processing pipelines and what implications there are for the user. Should you want to skip the technical details, head over to the <<benchmark,benchmark section>> below.
+
+Some of the details mentioned here were improved or introduced in the 0.21.0, but the compression ratio is equal between 0.20.0 and 0.21.0. Results are published against 0.22.0 which is the newest released version. 
+
+=== Compression algorithm
+
+In this release, we support the compression from the Facebook's great link:www.vldb.org/pvldb/vol8/p1816-teller.pdf[Gorilla paper] with small changes. For complete overview on how the compression method works I recommend reading the original paper. For short overview, the timestamps are compressed by calculating the delta of delta and writing them with variable length encoding while the data values are compressed by taking a XOR (eXclusive OR) between values (using the previous value as predictor for the next one) and storing that with variable length encoding also. We also slightly change the implementation details, as we need to use millisecond precision instead of the second precision used by Facebook.
+
+As we store values in a millisecond precision, it was important for us to increase the size of the block's maximum delta value. In Facebook's implementation they reserved 14 bits to represent the first delta from the block start (giving them 4 hour blocks with 1 second precision) where as we use 27 bits to get a maximum of 24 hour blocks with a millisecond precision. In 0.20.0/0.21.0 we however use 2 hour block sizes but with millisecond precision. We will revisit the block size requirements later.
+
+=== Insert pipeline
+
+We didn't change anything. When you issue a write to the Cassandra, we store it just like before to the `data` table. No restrictions on out-of-order times or changes to the persistence. Instead, we opted to do everything as an ETL (Extract Transform Load) job on the write path as described in the next section and merge results from both tables in the read path. 
```

Comment:
1st sentence is a bit short - what is "it" (probably the pipeline , but which / where ?)

Can you explain why we do it as we do it? (as opposed e.g. to compression on the fly)